pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,356 @@
+			Commitment: rpc.CommitmentFinalized,
+		})
+
+		if err != nil || accountInfo == nil || accountInfo.Value == nil {
+			return nil, fmt.Errorf("error fetching account info for PDA account: %s, error: %w", accountMeta.PublicKey.String(), err)
+		}
+
+		decoded, err := decodeBorshIntoType(accountInfo.GetBinary(), pda.InternalField.Type)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding Borsh data dynamically: %w", err)
+		}
+
+		value, err := GetValuesAtLocation(decoded, pda.InternalField.Location)
+		if err != nil {
+			return nil, fmt.Errorf("error getting value at location: %w", err)
+		}
+		if len(value) > 1 {
+			return nil, fmt.Errorf("multiple values found at location: %s", pda.InternalField.Location)
+		}
+
+		result = append(result, &solana.AccountMeta{
+			PublicKey:  solana.PublicKeyFromBytes(value[0]),
+			IsSigner:   accountMeta.IsSigner,
+			IsWritable: accountMeta.IsWritable,
+		})
+	}
+	return result, nil
+}
+
+func decodeBorshIntoType(data []byte, typ reflect.Type) (interface{}, error) {
+	// Ensure the type is a struct
+	if typ.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("provided type is not a struct: %s", typ.Kind())
+	}
+
+	// Create a new instance of the type
+	instance := reflect.New(typ).Interface()
+
+	// Decode using Borsh
+	err := ag_binary.NewBorshDecoder(data).Decode(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding Borsh data: %w", err)
+	}
+
+	// Return the underlying value (not a pointer)
+	return reflect.ValueOf(instance).Elem().Interface(), nil
+}
+
+// getSeedBytes extracts the seeds for the PDALookups.
+// It handles both AddressSeeds (which are public keys) and ValueSeeds (which are byte arrays from input args).
+func getSeedBytes(ctx context.Context, lookup PDALookups, args any, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([][]byte, error) {
+	var seedBytes [][]byte
+
+	for _, seed := range lookup.Seeds {
+		if lookupSeed, ok := seed.(AccountLookup); ok {
+			// Get value from a location (This doens't have to be an address, it can be any value)
+			bytes, err := GetValuesAtLocation(args, lookupSeed.Location)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+			seedBytes = append(seedBytes, bytes...)
+		} else {
+			// Get address seeds from the lookup
+			seedAddresses, err := GetAddresses(ctx, args, []Lookup{seed}, derivedTableMap, reader)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+
+			// Add each address seed as bytes
+			for _, address := range seedAddresses {
+				seedBytes = append(seedBytes, address.PublicKey.Bytes())
+			}
+		}
+	}
+
+	return seedBytes, nil
+}
+
+// generatePDAs generates program-derived addresses (PDAs) from public keys and seeds.
+func generatePDAs(publicKeys []*solana.AccountMeta, seeds [][]byte, lookup PDALookups) ([]*solana.AccountMeta, error) {
+	if len(seeds) > 1 && len(publicKeys) > 1 {
```

Comment:
Not really feedback but curious about this validation and this might just be my lack of knowledge on PDAs. So it's ok to process multiple publicKeys as long as we only have a single seed? And it's ok to process multiple seeds as long as we only have a single publicKey? Also on a separate note, are no seeds a valid scenario too?

Reply:
Honestly, I'm having trouble remembering why I added that validation in... I think it is okay to have multiple public keys and multiple seeds, each program ID would be used to generate a PDA with all the seeds and then all the addresses returned from the multiple PDAs would be returned by ChainWriter. I'm going to remove that validation. 

> Also on a separate note, are no seeds a valid scenario too?
I think technically no seeds is a valid scenario as long as a publicKey is supplied. The publicKey is also a seed but it just comes first and always corresponds to the overarching program / account address

Reply:
Noticed this validation is still here unless there was a reason for this. I also wanted to mention that we might want to validate the seeds here as well. According to the [docs](https://solana.com/developers/courses/native-onchain-development/program-derived-addresses) on PDAs:
```A PDA allows a maximum of 16 seeds, with each seed limited to 32 bytes in length```

Reply:
Good call out! Thought I removed that but guess I never got around to it. Also, I added validation for seed maximum [here](https://github.com/smartcontractkit/chainlink-solana/blob/135e6b9ecfaf7d055ea933685e8a3d274a448902/pkg/solana/chainwriter/lookups.go#L250-L252) and seed length [here](https://github.com/smartcontractkit/chainlink-solana/blob/135e6b9ecfaf7d055ea933685e8a3d274a448902/pkg/solana/chainwriter/lookups.go#L225-L230)